### PR TITLE
Supports multi-device model inference and wrapped forward functions

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes

Supports multi-device model inference and wrapped forward functions

### Reason for changes

Support tracing "bigscience/bloomz-560m" model from HF

### Related tickets

N/A

### Tests

test_no_self_forward,  test_multidevice_model

### CODE REVIEW

File: nncf/torch/nncf_network.py [hunk 1]
```diff
@@ -144,9 +144,10 @@ def get_original_forward(self) -> Callable:
         Returns the forward function of the original model, unmodified by NNCF. The returned function will
         have its 0-th implicit `self` argument bound to the model object.
         """
-        if self._original_instance_forward is not None:
-            return functools.partial(self._original_instance_forward, self._model_ref)
-        return functools.partial(self._original_unbound_forward, self._model_ref)
+        forward_parameters = self._forward_signature.parameters
+        if forward_parameters and next(iter(forward_parameters)) == "self":
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:148]: This is not a particularly reliable way of establishing whether the `self._forward_signature` is belonged to a bound or an unbound version of the method. `self` is not a reserved name, it's just a convention-level name for the 0-th argument. You can have:
```
import torch

class MyModule(torch.nn.Module):
    def forward(self, x, y):
        return x + y

def evil_forward(self, y):
    return self - y

module = MyModule()
module.forward = evil_forward

print(module.forward(20, 21))  # prints -1
```
Was this change strictly necessary to enable multi-device model inference? Maybe the case such as above should be added to tests?

File: nncf/torch/nncf_module_replacement.py [hunk 2]
```diff
@@ -186,6 +187,9 @@ def replace_modules_by_nncf_modules(
             else:
                 replaced_module = nncf_module_from(module)
 
+            device = get_model_device(module)
+            replaced_module.to(device)
+
             inter_dict[replaced_module] = scope_set
```
Comment (vshampor) [inline nncf/torch/nncf_module_replacement.py:193]: Better to get the model device before the replacer function, just in case the `nncf_module_from` does not return a copy of `module` but instead does in-place modifications.

File: nncf/torch/nncf_network.py (hunk 1 above)
Comment (alexsu52) [reply nncf/torch/nncf_network.py:148]: Could you suggest an approach to handle your case. And please provide the model in which this occurs?

File: nncf/torch/nncf_module_replacement.py (hunk 2 above)
Comment (alexsu52) [reply nncf/torch/nncf_module_replacement.py]: done

File: nncf/torch/dynamic_graph/io_handling.py [hunk 3]
```diff
@@ -218,7 +220,7 @@ def get_forward_inputs(self, device: str = None) -> Tuple[Tuple[torch.Tensor, ..
         return tuple(args_list), kwargs
 
 
-class ExactInputsInfo(ModelInputInfo):
+class ExampleInputInfo(ModelInputInfo):
     """
```
Comment (vshampor) [inline nncf/torch/dynamic_graph/io_handling.py:224]: You should still keep the `ExactInputsInfo` in the hierarchy and inherit `ExampleInputInfo` and `LoaderInputInfo` from it, since both those classes *are* providers of exact inputs to the model's forward.

File: nncf/torch/dynamic_graph/io_handling.py [hunk 4]
```diff
@@ -283,7 +287,7 @@ def from_nncf_config_dataloaders(cls, config: NNCFConfig) -> Optional["ExactInpu
                 wrapped_dataloader = wrap_dataloader_for_init(dataloader)
                 dataloader_output = next(iter(wrapped_dataloader))
                 args, kwargs = wrapped_dataloader.get_inputs(dataloader_output)
-                return ExactInputsInfo(args, kwargs)
+                return LoaderInputInfo(args, kwargs)
```
Comment (vshampor) [inline nncf/torch/dynamic_graph/io_handling.py:290]: `LoaderInputInfo` has no `__init__` right now. The `__init__` you want will be in the `ExactInputsInfo` class.

File: nncf/torch/dynamic_graph/io_handling.py [hunk 5]
```diff
@@ -227,7 +229,7 @@ def __init__(self, forward_args: Tuple, forward_kwargs: Dict):
         self._forward_args = forward_args
         self._forward_kwargs = forward_kwargs
```
Comment (vshampor) [inline nncf/torch/dynamic_graph/io_handling.py:230]: This is the `__init__` that should be on the `ExactInputsInfo` level of the hierarchy.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/torch/dynamic_graph/io_handling.py (hunk 3 above)
Comment (alexsu52) [reply nncf/torch/dynamic_graph/io_handling.py]: done

File: nncf/torch/dynamic_graph/io_handling.py (hunk 4 above)
Comment (alexsu52) [reply nncf/torch/dynamic_graph/io_handling.py:292]: done

File: nncf/torch/dynamic_graph/io_handling.py (hunk 5 above)
Comment (alexsu52) [reply nncf/torch/dynamic_graph/io_handling.py:230]: done

File: tests/torch/test_graph_building.py [hunk 6]
```diff
@@ -646,7 +647,8 @@ def test_filler_input_info_arg_generation(filler_gen_test_struct: FillerInputInf
     "input_info",
     [
         FillerInputInfo([FillerInputElement([1, 3, 3, 3])]),
-        ExactInputsInfo((torch.Tensor([1]), torch.Tensor([1])), {"a": torch.Tensor([1]), "b": torch.Tensor([1])}),
+        ExampleInputInfo((torch.Tensor([1]), torch.Tensor([1])), {"a": torch.Tensor([1]), "b": torch.Tensor([1])}),
+        LoaderInputInfo((torch.Tensor([1]), torch.Tensor([1])), {"a": torch.Tensor([1]), "b": torch.Tensor([1])}),
     ],
     ids=["filler", "exact"],
```
Comment (vshampor) [inline tests/torch/test_graph_building.py:653]: Add an ID for the new case and the remaining test should be passing.

**Review by vshampor — APPROVED** [review]

**Review by AlexanderDokuchaev — APPROVED** [review]
